### DISCUSSION TITLE
Image-Based Subtitle Support?

### DISCUSSION THREAD

**Discussion by dathbe** [jellyfin-archive/jellyfin-roku-legacy#987] [General]
Is there anything in the works to allow the use of PGSSUB, DVDSUB, and other image-based subtitle formats without requiring a transcode?  Most Blu-Rays don't come with text-based subtitles, so it would be really great to get support for image subtitles.  I have no idea how technologically difficult that may be, though.
Thanks.

**Comment by ghost** [comment]
I'm fairly new to all this, but my understanding is that image-based subtitles will always require transcoding because the server is having to merge the images together. The only way around that necessity would be to "burn in" the subtitles i.e. pre-render the subtitles in the video with Handbrake or something, but in that case the subtitles would always be present and you wouldn't be able to turn them on or off. Personally, I have had good results going to opensubtitles.org and downloading text-based subtitles for my movies, that could be an alternative that wouldn't require transcoding.

**Reply by dathbe** [reply]
Thanks for the reply.

I definitely don't have any technical expertise in that, but it seems like the Swiftfin iPhone app is able to display the image-based titles without transcoding.  VLC also does it which, while not a Jellyfin project, suggests it's technically possible.  

The problem I'm having with external subs is that it downmixes my movie sound.  See https://github.com/jellyfin/jellyfin-roku/issues/984.

**Comment by dathbe** [comment]
Update:

Workaround:

1) Add an external subtitle
2) use ffmpeg to mux the subtitle into the video file.  E.g. (mkv):

`ffmpeg -i withoutsubtitles.mkv -i withoutsubtitles.eng.srt -map 0 -map -0:s? -map 1 -map 0:s? -c copy -metadata:s:s:0 language=eng -disposition:s:0 default withsubtitles.mkv`

Explanation:

`-i` Input files
`-map 0` Add all streams from 1st input
`-map -0:s?` ...Except subtitles streams, if there are any (this just makes sure that the added subtitle is the first subtitle in the output)
`-map 1` Add all streams from the 2nd input
`-map 0:s?` Add all subtitle streams from the 1st input (this puts the subtitles from the original file back into the output after the new subtitle)
`-c copy` Do not transcode anything
`-metadata:s:s:0 language=eng -disposition:s:0 default` Set the language of the new subtitle to English, and set it to the default subtitle 
`withsubtitles.mkv` Output file

or (mp4)

`ffmpeg -i withoutsubtitles.mp4 -i withoutsubtitles.eng.srt -map 0:v -map 0:a -map 1 -map_metadata 0 -map_chapters 0 -c copy -c:s:0 mov_text -metadata:s:s:0 language=eng -disposition:s:0 default withsubtitles.mp4`

`-map 0:v` Add all _video_ streams from the 1st input (necessary because ffmpeg/mp4 do not always play nicely with some stream types)
`-map 0:a` Add all _audio_ streams from the 1st input
`-map_metadata 0` Add all metadata from the 1st input
`-map_chapters 0` Add all chapters from the 1st input
`-c:s:0 mov_text` Transcode the new subtitle to "mov_text" format (ffmpeg cannot insert SRT files into mp4; because this comes after `-c copy` it overrides for this stream)

This is a bit of a pain on anything more than 1 or 2 movies, so I've written a python script to look for where there are video and SRT files with the same name in the same folder and execute the correct ffmpeg command.  It then renames the old files to .old, and renames the new file to the original video file name so that Jellyfin will keep all information about the movie.
